File: apps/cli-rs/src/workspace/workspace_handle.rs
use crate::daemon::rpc::DomainRpcError;
use crate::workspace::{
    file_service::FileService, git_service::GitService, terminal::TerminalManager, types::*,
};
use std::collections::HashMap;
use std::sync::Arc;
use std::sync::RwLock;

/// Workspace-scoped facade over file, git, and terminal services.
pub struct WorkspaceHandle<'a> {
    pub(super) workspace: Workspace,
    pub(super) workspaces: &'a RwLock<HashMap<String, Workspace>>,
    pub(super) files: &'a FileService,
    pub(super) gits: &'a GitService,
    pub(super) terminals: &'a TerminalManager,
}

impl WorkspaceHandle<'_> {
    pub fn set_pull_request(&self, pr: Option<WorkspacePullRequest>) -> Result<(), DomainRpcError> {
        let mut workspaces = self.workspaces.write().unwrap();
        let workspace = workspaces.get_mut(&self.workspace.id).ok_or_else(|| {
            DomainRpcError::not_found(format!("workspace not found: {}", self.workspace.id))
        })?;
        workspace.pull_request = pr;
        Ok(())
    }

    pub fn file_list(&self, path: &str, recursive: bool) -> Result<Vec<FileEntry>, DomainRpcError> {
        self.files.list(&self.workspace.path, path, recursive)
    }

    pub fn file_stat(&self, path: &str) -> Result<FileEntry, DomainRpcError> {
        self.files.stat(&self.workspace.path, path)
    }

    pub fn file_read(&self, path: &str) -> Result<String, DomainRpcError> {
        self.files.read(&self.workspace.path, path)
    }

    pub fn file_write(
        &self,
        path: &str,
        content: &str,
        mode: u32,
    ) -> Result<usize, DomainRpcError> {
        self.files.write(&self.workspace.path, path, content, mode)
    }

    pub fn file_delete(&self, path: &str, recursive: bool) -> Result<(), DomainRpcError> {
        self.files.delete(&self.workspace.path, path, recursive)
    }

    pub fn file_move(&self, from: &str, to: &str) -> Result<(), DomainRpcError> {
        self.files.move_path(&self.workspace.path, from, to)
    }

    pub fn file_mkdir(&self, path: &str, parents: bool, mode: u32) -> Result<(), DomainRpcError> {
        self.files.mkdir(&self.workspace.path, path, parents, mode)
    }

    pub fn file_read_diff(&self, path: &str) -> Result<GitDiffContent, DomainRpcError> {
        self.files.read_diff(&self.workspace.path, path)
    }

    pub fn git_status(&self) -> Result<GitStatusResponse, DomainRpcError> {
        self.gits.status(&self.workspace.path)
    }

    pub fn git_inspect(&self) -> Result<GitInspectResult, DomainRpcError> {
        self.gits.inspect(&self.workspace.path)
    }

    pub fn git_list_changes(&self) -> Result<GitChangesBySection, DomainRpcError> {
        self.gits.list_changes(&self.workspace.path)
    }

    pub fn git_track(&self, paths: &[String]) -> Result<(), DomainRpcError> {
        self.gits.track_changes(&self.workspace.path, paths)
    }

    pub fn git_unstage(&self, paths: &[String]) -> Result<(), DomainRpcError> {
        self.gits.unstage_changes(&self.workspace.path, paths)
    }

    pub fn git_revert(&self, paths: &[String]) -> Result<(), DomainRpcError> {
        self.gits.revert_changes(&self.workspace.path, paths)
    }

    pub fn git_commit(
        &self,
        message: &str,
        amend: bool,
        signoff: bool,
    ) -> Result<String, DomainRpcError> {
        self.gits
            .commit_changes(&self.workspace.path, message, amend, signoff)
    }

    pub fn current_branch(&self) -> Result<String, DomainRpcError> {
        self.gits.current_branch(&self.workspace.path)
    }

    pub fn git_branch_status(&self) -> Result<GitBranchStatus, DomainRpcError> {
        self.gits.branch_status(&self.workspace.path)
    }

    pub fn git_branch_pr(
        &self,
        branch: &str,
    ) -> Result<GitBranchPullRequestStatus, DomainRpcError> {
        self.gits.branch_pull_request(&self.workspace.path, branch)
    }

    pub fn git_commits_to_target(
        &self,
        target: &str,
    ) -> Result<GitCommitComparison, DomainRpcError> {
        self.gits
            .list_commits_to_target(&self.workspace.path, target)
    }

    pub fn git_branch_diff_summary(
        &self,
        target: &str,
    ) -> Result<GitBranchDiffSummary, DomainRpcError> {
        self.gits.branch_diff_summary(&self.workspace.path, target)
    }

    pub fn git_commit_diff(
        &self,
        hash: &str,
        path: &str,
    ) -> Result<GitDiffContent, DomainRpcError> {
        self.gits.read_commit_diff(&self.workspace.path, hash, path)
    }

    pub fn git_branch_diff(
        &self,
        target: &str,
        path: &str,
    ) -> Result<GitDiffContent, DomainRpcError> {
        self.gits
            .read_branch_comparison_diff(&self.workspace.path, target, path)
    }

    pub fn git_branches(&self) -> Result<GitBranchList, DomainRpcError> {
        self.gits.list_branches(&self.workspace.path)
    }

    pub fn git_push(&self) -> Result<String, DomainRpcError> {
        self.gits.push_branch(&self.workspace.path)
    }

    pub fn git_publish(&self) -> Result<String, DomainRpcError> {
        self.gits.publish_branch(&self.workspace.path)
    }

    pub fn git_rename_branch(&self, next_branch: &str) -> Result<(), DomainRpcError> {
        self.gits.rename_branch(&self.workspace.path, next_branch)
    }

    pub fn git_remove_branch(&self, branch: &str, force: bool) -> Result<(), DomainRpcError> {
        self.gits.remove_branch(&self.workspace.path, branch, force)
    }

    pub fn git_pr_merge(
        &self,
        pr_number: i64,
        method: &str,
        delete_branch: bool,
    ) -> Result<String, DomainRpcError> {
        self.gits
            .merge_pull_request(&self.workspace.path, pr_number, method, delete_branch)
    }

    pub fn git_pr_close(&self, pr_number: i64) -> Result<String, DomainRpcError> {
        self.gits
            .close_pull_request(&self.workspace.path, pr_number)
    }

    pub fn git_worktree_create(
        &self,
        branch: &str,
        worktree_path: &str,
        create_branch: bool,
        from_ref: &str,
    ) -> Result<(), DomainRpcError> {
        self.gits.create_worktree(
            &self.workspace.path,
            branch,
            worktree_path,
            create_branch,
            from_ref,
        )
    }

    pub fn git_worktree_remove(
        &self,
        worktree_path: &str,
        force: bool,
    ) -> Result<(), DomainRpcError> {
        let main_worktree_path = self.gits.main_worktree_path(&self.workspace.path)?;
        self.gits
            .remove_worktree(&main_worktree_path, worktree_path, force)
    }

    pub fn git_author_name(&self) -> Result<String, DomainRpcError> {
        self.gits.author_name(&self.workspace.path)
    }

    pub fn terminal_start(
        &self,
        req: &TerminalStartRequest,
    ) -> Result<TerminalStartResponse, DomainRpcError> {
        self.terminals.start(&self.workspace.path, req)
    }

    pub fn terminal_send(
        &self,
        req: &TerminalSendRequest,
    ) -> Result<TerminalSendResponse, DomainRpcError> {
        self.terminals.send(req)
    }

    pub fn terminal_read(
        &self,
        req: &TerminalReadRequest,
    ) -> Result<TerminalReadResponse, DomainRpcError> {
        self.terminals.read(req)
    }

    pub fn terminal_stop(
        &self,
        req: &TerminalStopRequest,
    ) -> Result<TerminalStopResponse, DomainRpcError> {
        self.terminals.stop(req)
    }

    pub fn terminal_list_sessions(&self) -> Vec<TerminalSessionSummary> {
        self.terminals.list_sessions(&TerminalListSessionsRequest {
            workspace_id: Some(self.workspace.id.clone()),
        })
    }

    pub fn terminal_resize(
        &self,
        req: &TerminalResizeRequest,
    ) -> Result<TerminalResizeResponse, DomainRpcError> {
        self.terminals.resize(req)
    }

    pub fn terminal_subscribe_output(
        &self,
        session_id: &str,
        sink: Arc<
            tokio::sync::Mutex<
                futures_util::stream::SplitSink<
                    axum::extract::ws::WebSocket,
                    axum::extract::ws::Message,
                >,
            >,
        >,
    ) -> Result<(), DomainRpcError> {
        self.terminals.subscribe_output(session_id, sink)
    }
}

File: apps/cli/internal/workspace/workspace_handle.go
package workspace

import (
	"context"

	"yishan/apps/cli/internal/workspace/terminal"
)

type WorkspaceHandle struct {
	workspace Workspace
	files     *FileService
	gits      *GitService
	terminals *terminal.Manager
}

func (h WorkspaceHandle) Workspace() Workspace {
	return h.workspace
}

func (h WorkspaceHandle) FileList(path string, recursive bool) ([]FileEntry, error) {
	return h.files.List(h.workspace.Path, path, recursive)
}

func (h WorkspaceHandle) FileSearch(query string, limit int) ([]FileSearchResult, error) {
	return h.files.Search(h.workspace.Path, query, limit)
}

func (h WorkspaceHandle) FileStat(path string) (FileEntry, error) {
	return h.files.Stat(h.workspace.Path, path)
}

func (h WorkspaceHandle) FileRead(path string) (string, error) {
	return h.files.Read(h.workspace.Path, path)
}

func (h WorkspaceHandle) FileWrite(path string, content string, mode uint32) (int, error) {
	return h.files.Write(h.workspace.Path, path, content, mode)
}

func (h WorkspaceHandle) FileDelete(path string, recursive bool) error {
	return h.files.Delete(h.workspace.Path, path, recursive)
}

func (h WorkspaceHandle) FileMove(fromPath string, toPath string) error {
	return h.files.Move(h.workspace.Path, fromPath, toPath)
}

func (h WorkspaceHandle) FileMkdir(path string, parents bool, mode uint32) error {
	return h.files.Mkdir(h.workspace.Path, path, parents, mode)
}

func (h WorkspaceHandle) FileReadDiff(ctx context.Context, path string) (GitDiffContent, error) {
	return h.files.ReadDiff(ctx, h.workspace.Path, path)
}

func (h WorkspaceHandle) TerminalStart(ctx context.Context, req TerminalStartRequest) (TerminalStartResponse, error) {
	return h.terminals.Start(ctx, h.workspace.Path, req)
}

func (h WorkspaceHandle) GitStatus(ctx context.Context) (GitStatusResponse, error) {
	return h.gits.Status(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitListChanges(ctx context.Context) (GitChangesBySection, error) {
	return h.gits.ListChanges(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitTrackChanges(ctx context.Context, paths []string) error {
	return h.gits.TrackChanges(ctx, h.workspace.Path, paths)
}

func (h WorkspaceHandle) GitUnstageChanges(ctx context.Context, paths []string) error {
	return h.gits.UnstageChanges(ctx, h.workspace.Path, paths)
}

func (h WorkspaceHandle) GitRevertChanges(ctx context.Context, paths []string) error {
	return h.gits.RevertChanges(ctx, h.workspace.Path, paths)
}

func (h WorkspaceHandle) GitCommitChanges(ctx context.Context, message string, amend bool, signoff bool) (string, error) {
	return h.gits.CommitChanges(ctx, h.workspace.Path, message, amend, signoff)
}

func (h WorkspaceHandle) GitBranchStatus(ctx context.Context) (GitBranchStatus, error) {
	return h.gits.BranchStatus(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitBranchPullRequest(ctx context.Context, branch string) (GitBranchPullRequestStatus, error) {
	return h.gits.BranchPullRequest(ctx, h.workspace.Path, branch)
}

func (h WorkspaceHandle) RefreshGitBranchPullRequest(ctx context.Context, branch string) (GitBranchPullRequestStatus, error) {
	return h.gits.RefreshBranchPullRequest(ctx, h.workspace.Path, branch)
}

func (h WorkspaceHandle) GitCurrentBranch(ctx context.Context) (string, error) {
	return h.gits.CurrentBranch(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitBranchPullRequestLite(ctx context.Context, branch string) (GitBranchPullRequestStatus, error) {
	return h.gits.BranchPullRequestLite(ctx, h.workspace.Path, branch)
}

func (h WorkspaceHandle) GitBranchPullRequestWithDetails(ctx context.Context, branch string) (GitBranchPullRequestStatus, error) {
	return h.gits.BranchPullRequestWithDetails(ctx, h.workspace.Path, branch)
}

func (h WorkspaceHandle) GitListCommitsToTarget(ctx context.Context, targetBranch string) (GitCommitComparison, error) {
	return h.gits.ListCommitsToTarget(ctx, h.workspace.Path, targetBranch)
}

func (h WorkspaceHandle) GitBranchDiffSummary(ctx context.Context, targetBranch string) (GitBranchDiffSummary, error) {
	return h.gits.BranchDiffSummary(ctx, h.workspace.Path, targetBranch)
}

func (h WorkspaceHandle) GitReadCommitDiff(ctx context.Context, commitHash string, path string) (GitDiffContent, error) {
	return h.gits.ReadCommitDiff(ctx, h.workspace.Path, commitHash, path)
}

func (h WorkspaceHandle) GitReadBranchComparisonDiff(ctx context.Context, targetBranch string, path string) (GitDiffContent, error) {
	return h.gits.ReadBranchComparisonDiff(ctx, h.workspace.Path, targetBranch, path)
}

func (h WorkspaceHandle) GitListBranches(ctx context.Context) (GitBranchList, error) {
	return h.gits.ListBranches(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitPushBranch(ctx context.Context) (string, error) {
	return h.gits.PushBranch(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitPublishBranch(ctx context.Context) (string, error) {
	return h.gits.PublishBranch(ctx, h.workspace.Path)
}

func (h WorkspaceHandle) GitRenameBranch(ctx context.Context, nextBranch string) error {
	return h.gits.RenameBranch(ctx, h.workspace.Path, nextBranch)
}

func (h WorkspaceHandle) GitRemoveBranch(ctx context.Context, branch string, force bool) error {
	return h.gits.RemoveBranch(ctx, h.workspace.Path, branch, force)
}

func (h WorkspaceHandle) GitPrMerge(ctx context.Context, prNumber int, method string, deleteBranch bool) (string, error) {
	return h.gits.MergePullRequest(ctx, h.workspace.Path, prNumber, method, deleteBranch)
}

func (h WorkspaceHandle) GitPrClose(ctx context.Context, prNumber int) (string, error) {
	return h.gits.ClosePullRequest(ctx, h.workspace.Path, prNumber)
}

func (h WorkspaceHandle) GitCreateWorktree(ctx context.Context, branch string, worktreePath string, createBranch bool, fromRef string) error {
	return h.gits.CreateWorktree(ctx, h.workspace.Path, branch, worktreePath, createBranch, fromRef)
}

func (h WorkspaceHandle) GitRemoveWorktree(ctx context.Context, worktreePath string, force bool) error {
	return h.gits.RemoveWorktree(ctx, h.workspace.Path, worktreePath, force)
}

func (h WorkspaceHandle) GitAuthorName(ctx context.Context) (string, error) {
	return h.gits.AuthorName(ctx, h.workspace.Path)
}
